sd_bandits/obp_extensions/dataset.py

```diff
@@ -188,4 +327,142 @@ def obtain_batch_bandit_feedback(
+                    Array of actions presented.
+                reward : np.array (n_rounds,)
+                    Array of rewards returned.
+                position : np.array (n_rounds,)
+                    Array of the position of each action.
+                context : np.array (n_rounds, 97)
+                    Array of the user context for each round.
+                action_context : np.array (n_rounds, 97)
+                    Array of the action context for each round.
+                pscore: NoneType
+                    N/A.
+                n_rounds : int
+                    Number of users per batch * number of batches * number of items observed for that
+                    user's session.
+                n_actions : int
+                    Number of available items (always 862).
+                users : np.array (n_rounds * users_per_round,)
+                    The users that were generated for this simulation.
+
+        """
+        policy.batch_size = np.inf
+        rng = np.random.default_rng(seed)
+        user_indices = rng.choice(
+            range(len(self.user_features)),
+            size=users_per_batch * n_batches,
+            replace=True,
+        )
+
+        all_probabilities = []
+        all_item_indices = []
+
+        relevant_user_features = self.user_features[user_indices, :]
+
+        actions = []
+        rewards = []
+        positions = []
+        context = []
+        selected_actions = []
+
+        for i, user_idx in tqdm(
+            enumerate(user_indices),
+            desc="Simulating online learning",
+            total=len(user_indices),
+        ):
+            if policy.policy_type == "contextfree":
+                item_indices = policy.select_action()
+            else:
+                item_indices = policy.select_action(self.user_features[user_idx])
+            all_item_indices.append(item_indices)
+            this_user_features = self.user_features[user_idx, :]
+            these_items_features = self.playlist_features[item_indices, :]
+            # calculate raw user-playlist affinities
+            raw_scores = these_items_features @ this_user_features
+            probabilities = expit(raw_scores)
+
+            # observe all rewards
+            rewards_uncascaded = rng.binomial(n=1, p=probabilities)
+            positions_to_observe = self._get_positions_to_observe(
+                rewards_uncascaded, cascade, cascade_at
+            )
+
+            for position in range(positions_to_observe):
+                is_last_position = position == positions_to_observe - 1
+                is_end_of_batch = (i + 1) % users_per_batch == 0
+                if is_last_position and is_end_of_batch:
+                    # sneakily trigger update
+                    policy.batch_size = policy.n_trial + 1
+                reward = rewards_uncascaded[position]
+                action = item_indices[position]
+                policy.update_params(action=action, reward=reward)
```

Comment:
we can also add a `segment` parameter to `update_params()` and just have it set to None if it is not a segment-based policy